libraries/chain/fork_database.cpp

```diff
@@ -21,79 +20,154 @@ namespace eosio::chain {
     * Version 1: initial version of the new refactored fork database portable format
     */
 
+   // call while holding fork database lock
+   struct block_state_accessor {
+      static bool is_valid(const block_state& bs) {
+         return bs.is_valid();
+      }
+      static void set_valid(block_state& bs, bool v) {
+         bs.validated = v;
+      }
+      static uint32_t last_final_block_num(const block_state& bs) {
+         return bs.updated_core.last_final_block_num;
+      }
+      static uint32_t final_on_strong_qc_block_num(const block_state& bs) {
+         return bs.updated_core.final_on_strong_qc_block_num;
+      }
+      static uint32_t lastest_qc_claim_block_num(const block_state& bs) {
+         return bs.updated_core.latest_qc_claim_block_num;
+      }
+      static bool qc_claim_update_needed(block_state& bs, const qc_claim_t& most_recent_ancestor_with_qc) {
+         return bs.most_recent_ancestor_with_qc < most_recent_ancestor_with_qc;
+      }
+      static bool qc_claim_update_needed(block_state& bs, const block_state& prev) {
+         return bs.most_recent_ancestor_with_qc < prev.most_recent_ancestor_with_qc;
+      }
+      static void update_best_qc(block_state& bs, const qc_claim_t& most_recent_ancestor_with_qc) {
+         assert(bs.most_recent_ancestor_with_qc < most_recent_ancestor_with_qc);
```

Comment:
Is this assert always true when this method is called? You can also just check if `<` then return.

Reply:
Yes, the expectation is that `qc_claim_update_needed` is called first. This is internal to `fork_database.cpp`, so used an `assert` instead of an if-check.